File: apps/api/src/services/llm.ts
import { config } from "../config.js";
import type { OutlineItem } from "../db.js";
import type { Layout, SlideIn } from "../repo.js";

const ALLOWED = new Set(["title", "section", "bullets", "two_column"]);

function expandSystem(): string {
  return `You expand messy meeting notes into a speakable narrative for a presentation.
Return ONLY valid JSON (no markdown fences):
{
  "title": "deck title",
  "expanded_notes": "full markdown string",
  "outline": [{"title": "chapter", "intent": "one-line purpose"}]
}

Hard rules (expand-narrative):
- Keep names, numbers, product names, metaphors from the notes. Do NOT invent facts, metrics, quotes, or events.
- Add connective logic and speaker-ready phrasing only ("so what?", transitions). Prefer one clear narrative arc.
- expanded_notes: markdown with elevator summary then H2 sections; speakable, dense bullets; cull filler and repetition.
- outline: TARGET ${config.maxOutlineChapters} chapters max (prefer 4–${config.maxOutlineChapters}). Merge semantically similar themes into one chapter (e.g. brand perception + entry points; talent + creativity).
- Order for speaking: open → body → close. Do not mirror note-taking order if it is fragmented.
`;
}

function slidesSystem(): string {
  return `You turn an EXPANDED markdown narrative (+ optional outline) into presentation slides.
Return ONLY valid JSON (no markdown fences):
{"title": "...", "slides": [{"layout": "...", "title": "...", "body": {}}]}

Hard rules:
- Build slides from the expanded narrative only.
- TOTAL slides MUST be ≤ ${config.maxSlides} (including cover and closing). Prefer ${Math.min(10, config.maxSlides)}–${config.maxSlides}.
- Allowed layouts: title, section, bullets, two_column
- body title/section: {"subtitle": "..."}
- body bullets: {"bullets": ["...", "..."]} — use 5–8 dense points per content slide (conclusion + evidence where possible)
- body two_column: {"left": [...], "right": [...], "left_title": "", "right_title": ""} — prefer for contrasts
- Follow outline order when provided, but MERGE near-duplicate chapters. Do NOT emit one section + many pages per chapter.
- Use at most 2–3 section dividers for the whole deck. Prefer high-density bullets/two_column over many thin slides.
- Cover first; short closing last; no emoji; do not invent facts beyond the expanded text.
`;
}

function stripFences(text: string): string {
  let t = text.trim();
  if (t.startsWith("```")) {
    t = t.replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/, "");
  }
  return t.trim();
}

function httpError(statusCode: number, detail: string): Error {
  const err = new Error(detail);
  (err as Error & { statusCode: number }).statusCode = statusCode;
  return err;
}

async function chat(
  messages: { role: string; content: string }[],
  temperature = 0.3,
): Promise<unknown> {
  if (!config.llmApiKey) {
    throw httpError(503, "LLM_API_KEY is not configured");
  }
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), config.llmTimeoutMs);
  try {
    const res = await fetch(`${config.llmBaseUrl}/chat/completions`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${config.llmApiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: config.llmModel,
        messages,
        temperature,
      }),
      signal: controller.signal,
    });
    if (!res.ok) {
      throw httpError(502, `LLM request failed with status ${res.status}`);
    }
    return res.json();
  } catch (e) {
    if (e instanceof Error && e.name === "AbortError") {
      throw httpError(504, "LLM request timed out");
    }
    throw e;
  } finally {
    clearTimeout(timeoutId);
  }
}

function parseJsonContent(data: unknown): Record<string, unknown> {
  const content = (data as { choices?: { message?: { content?: string } }[] })
    ?.choices?.[0]?.message?.content;
  if (typeof content !== "string" || !content.trim()) {
    throw httpError(502, "LLM returned empty response");
  }
  try {
    const parsed = JSON.parse(stripFences(content));
    if (!parsed || typeof parsed !== "object") {
      throw httpError(502, "LLM returned invalid JSON");
    }
    return parsed as Record<string, unknown>;
  } catch (e) {
    if ((e as Error & { statusCode?: number }).statusCode) throw e;
    throw httpError(502, "LLM returned invalid JSON");
  }
}

function normalizeBody(layout: string, body: unknown): Record<string, unknown> {
  const b = body && typeof body === "object" ? (body as Record<string, unknown>) : {};
  if (layout === "title" || layout === "section") {
    return { subtitle: String(b.subtitle ?? "") };
  }
  if (layout === "two_column") {
    const left = Array.isArray(b.left) ? b.left.map(String) : [];
    const right = Array.isArray(b.right) ? b.right.map(String) : [];
    return {
      left,
      right,
      left_title: String(b.left_title ?? ""),
      right_title: String(b.right_title ?? ""),
    };
  }
  const bullets = Array.isArray(b.bullets) ? b.bullets.map(String) : [];
  return { bullets };
}

function normalizeSlides(raw: unknown): SlideIn[] {
  if (!Array.isArray(raw)) return [];
  const slides: SlideIn[] = [];
  raw.forEach((item, i) => {
    if (!item || typeof item !== "object") return;
    const obj = item as Record<string, unknown>;
    let layout = String(obj.layout || "bullets");
    if (!ALLOWED.has(layout)) layout = "bullets";
    slides.push({
      position: i,
      layout: layout as Layout,
      title: String(obj.title ?? ""),
      body: normalizeBody(layout, obj.body),
    });
  });
  return slides;
}

function normalizeOutline(raw: unknown): OutlineItem[] {
  if (!Array.isArray(raw)) return [];
  return raw
    .filter((x) => x && typeof x === "object")
    .map((x) => ({
      title: String((x as OutlineItem).title || "").trim(),
      intent: String((x as OutlineItem).intent || "").trim(),
    }))
    .filter((x) => x.title)
    .slice(0, config.maxOutlineChapters);
}

export type ExpandLLMResult = {
  title: string;
  expanded_notes: string;
  outline: OutlineItem[];
};

export type StructureLLMResult = {
  title: string;
  slides: SlideIn[];
};

export async function expandNotes(
  rawNotes: string,
  title?: string | null,
): Promise<ExpandLLMResult> {
  const parts = [`Meeting notes:\n${rawNotes}`];
  if (title) parts.unshift(`Preferred deck title: ${title}`);
  const data = await chat(
    [
      { role: "system", content: expandSystem() },
      { role: "user", content: parts.join("\n\n") },
    ],
    0.35,
  );
  const parsed = parseJsonContent(data);
  const deckTitle =
    (title || "").trim() ||
    String(parsed.title || "Untitled").trim() ||
    "Untitled";
  const expanded = String(parsed.expanded_notes || "").trim();
  if (!expanded) throw httpError(502, "LLM returned empty expanded_notes");
  return {
    title: deckTitle,
    expanded_notes: expanded,
    outline: normalizeOutline(parsed.outline),
  };
}

export async function slidesFromExpanded(
  expandedNotes: string,
  opts: { title?: string | null; outline?: OutlineItem[] } = {},
): Promise<StructureLLMResult> {
  if (!expandedNotes.trim()) {
    throw httpError(400, "expanded_notes is required before generating slides");
  }
  const parts = [`Expanded narrative:\n${expandedNotes}`];
  if (opts.outline?.length) {
    const lines = opts.outline
      .map((o) => `- ${o.title}${o.intent ? ` — ${o.intent}` : ""}`)
      .join("\n");
    parts.unshift(`Confirmed outline (follow this order; merge similar):\n${lines}`);
  }
  if (opts.title) parts.unshift(`Preferred deck title: ${opts.title}`);
  const data = await chat(
    [
      { role: "system", content: slidesSystem() },
      { role: "user", content: parts.join("\n\n") },
    ],
    0.25,
  );
  const parsed = parseJsonContent(data);
  const deckTitle =
    (opts.title || "").trim() ||
    String(parsed.title || "Untitled").trim() ||
    "Untitled";
  return {
    title: deckTitle,
    slides: normalizeSlides(parsed.slides),
  };
}

File: backend/app/services/llm.py
import json
import re

import httpx
from fastapi import HTTPException

from app.config import get_settings
from app.schemas import SlideIn, StructureLLMResult

ALLOWED_LAYOUTS = frozenset({"title", "section", "bullets", "two_column"})

SYSTEM_PROMPT = """You structure messy meeting notes into presentation slides.
Return ONLY valid JSON (no markdown fences) with this shape:
{"title": "...", "slides": [{"layout": "...", "title": "...", "body": {}}]}

Rules:
- Allowed layouts only: title, section, bullets, two_column
- body for title/section: {"subtitle": "..."}
- body for bullets: {"bullets": ["...", "..."]}
- body for two_column: {"left": [...], "right": [...], "left_title": "", "right_title": ""}
- Prefer cover (title) → section dividers → bullet pages; optional closing summary
- One focus per slide; short bullet phrases
- Keep key terms, names, product names from the notes
- Do not invent facts not present in the notes
"""


def _strip_fences(text: str) -> str:
    text = text.strip()
    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?\s*", "", text, count=1, flags=re.IGNORECASE)
        text = re.sub(r"\s*```$", "", text, count=1)
    return text.strip()


def _normalize_body(layout: str, body: object) -> dict:
    if not isinstance(body, dict):
        body = {}
    if layout == "title" or layout == "section":
        return {"subtitle": str(body.get("subtitle", "") or "")}
    if layout == "two_column":
        left = body.get("left", [])
        right = body.get("right", [])
        if not isinstance(left, list):
            left = []
        if not isinstance(right, list):
            right = []
        return {
            "left": [str(x) for x in left],
            "right": [str(x) for x in right],
            "left_title": str(body.get("left_title", "") or ""),
            "right_title": str(body.get("right_title", "") or ""),
        }
    # bullets (default)
    bullets = body.get("bullets", [])
    if not isinstance(bullets, list):
        bullets = []
    return {"bullets": [str(b) for b in bullets]}


def _normalize_slides(raw_slides: object) -> list[SlideIn]:
    if not isinstance(raw_slides, list):
        raw_slides = []
    slides: list[SlideIn] = []
    for i, item in enumerate(raw_slides):
        if not isinstance(item, dict):
            continue
        layout = item.get("layout", "bullets")
        if layout not in ALLOWED_LAYOUTS:
            layout = "bullets"
        body = _normalize_body(layout, item.get("body"))
        slides.append(
            SlideIn(
                position=i,
                layout=layout,
                title=str(item.get("title", "") or ""),
                body=body,
            )
        )
    return slides


def structure_notes(raw_notes: str, title: str | None = None) -> StructureLLMResult:
    settings = get_settings()
    if not settings.llm_api_key:
        raise HTTPException(status_code=503, detail="LLM_API_KEY is not configured")

    base = settings.llm_base_url.rstrip("/")
    url = f"{base}/chat/completions"
    user_parts = [f"Meeting notes:\n{raw_notes}"]
    if title:
        user_parts.insert(0, f"Preferred deck title: {title}")

    payload = {
        "model": settings.llm_model,
        "messages": [
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": "\n\n".join(user_parts)},
        ],
        "temperature": 0.2,
    }

    try:
        with httpx.Client(timeout=60.0) as client:
            resp = client.post(
                url,
                headers={
                    "Authorization": f"Bearer {settings.llm_api_key}",
                    "Content-Type": "application/json",
                },
                json=payload,
            )
            resp.raise_for_status()
            data = resp.json()
    except httpx.HTTPStatusError as exc:
        raise HTTPException(
            status_code=502,
            detail=f"LLM request failed with status {exc.response.status_code}",
        ) from exc
    except httpx.HTTPError as exc:
        raise HTTPException(status_code=502, detail="LLM request failed") from exc

    try:
        content = data["choices"][0]["message"]["content"]
    except (KeyError, IndexError, TypeError) as exc:
        raise HTTPException(status_code=502, detail="LLM returned empty response") from exc

    if not isinstance(content, str) or not content.strip():
        raise HTTPException(status_code=502, detail="LLM returned empty response")

    try:
        parsed = json.loads(_strip_fences(content))
    except json.JSONDecodeError as exc:
        raise HTTPException(status_code=502, detail="LLM returned invalid JSON") from exc

    if not isinstance(parsed, dict):
        raise HTTPException(status_code=502, detail="LLM returned invalid JSON")

    deck_title = (title or "").strip() or str(parsed.get("title") or "Untitled").strip() or "Untitled"
    slides = _normalize_slides(parsed.get("slides"))
    return StructureLLMResult(title=deck_title, slides=slides)
